src/libcore/slice/mod.rs

```diff
@@ -1402,6 +1402,172 @@ impl<T> [T] {
+            //     +---+---+---+---+---+---+
+            //     | 0 | 1 | 1 | 2 | 3 | 3 |
+            //     +---+---+---+---+---+---+
+            //               w
+            //
+            // Comparing self[r] against self[w-1], this value is a duplicate,
+            // so we increment `r` but leave everything else unchanged:
+            //
+            //                   r
+            //     +---+---+---+---+---+---+
+            //     | 0 | 1 | 1 | 2 | 3 | 3 |
+            //     +---+---+---+---+---+---+
+            //               w
+            //
+            // Comparing self[r] against self[w-1], this is not a duplicate,
+            // so swap self[r] and self[w] and advance r and w:
+            //
+            //                       r
+            //     +---+---+---+---+---+---+
+            //     | 0 | 1 | 2 | 1 | 3 | 3 |
+            //     +---+---+---+---+---+---+
+            //                   w
+            //
+            // Not a duplicate, repeat:
+            //
+            //                           r
+            //     +---+---+---+---+---+---+
+            //     | 0 | 1 | 2 | 3 | 1 | 3 |
+            //     +---+---+---+---+---+---+
+            //                       w
+            //
+            // Duplicate, advance r. End of slice. Split at w.
+
+            let ln = self.len();
+            if ln <= 1 {
+                return (self, &mut [])
+            }
+
+            // Avoid bounds checks by using raw pointers.
+            let p = self.as_mut_ptr();
+            let mut r: usize = 1;
+            let mut w: usize = 1;
+
+            while r < ln {
+                let p_r = p.add(r);
+                let p_wm1 = p.add(w - 1);
+                if !same_bucket(&mut *p_r, &mut *p_wm1) {
+                    if r != w {
+                        let p_w = p_wm1.offset(1);
+                        mem::swap(&mut *p_r, &mut *p_w);
+                    }
+                    w += 1;
+                }
+                r += 1;
+            }
+
+            self.split_at_mut(w)
+        }
+    }
+
+    /// Moves all but the first of consecutive elements to the end of the slice that resolve
+    /// to the same key. Returns two slices, the first contains the deduplicated elements,
+    /// the second contains all the duplicates unordered.
+    ///
+    /// If the slice is sorted, the first slice contains no duplicates.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// #![feature(slice_retain_dedup)]
+    ///
+    /// let mut slice = [10, 20, 21, 30, 30, 20];
+    ///
+    /// let (dedup, duplicates) = slice.retain_dedup_by_key(|i| *i / 10);
+    ///
+    /// assert_eq!(dedup, [10, 20, 30, 20]);
+    /// assert_eq!(duplicates, [30, 21]);
+    /// ```
+    #[unstable(feature = "slice_retain_dedup", issue = "0")]
+    #[inline]
+    pub fn retain_dedup_by_key<K, F>(&mut self, mut key: F) -> (&mut [T], &mut [T])
+        where F: FnMut(&T) -> K,
+              K: Ord,
```

Comment:
[`Vec::dedup_by_key`](https://doc.rust-lang.org/nightly/alloc/vec/struct.Vec.html#method.dedup_by_key) has `F: FnMut(&mut T) -> K` and `K: PartialEq`. Is there a reason to differ here?